[ADD] partner_invoicing_mode_ten_days

rousseldenis opened the pull request:
Depends on:

- #1515

rousseldenis committed:
[ADD] partner_invoicing_mode_ten_days

rousseldenis committed:
[DONT MERGE] test-requirements.txt

rousseldenis committed:
[IMP] partner_invoicing_mode_ten_days: Allow uninvoiced periods to be relaunched

The former process was too restrictive. So, allow to launch the cron at every time. The pivot
date (10th/20th/last) is now used to filter the last_execution.

sbejaoui commented on partner_invoicing_mode_ten_days/models/res_partner.py:
```diff
@@ -0,0 +1,40 @@
+            lambda partner: partner.invoicing_mode == "ten_days"
+        )
+        if ten_days_partners:
+            rows = [10, 20, (today + relativedelta(day=31)).day]
+            try:
+                next_day = next(row for row in rows if today.day < row)
+                next_invoice_date = today.replace(day=next_day)
+            except StopIteration:
+                next_day = rows[0]
+                next_invoice_date = (today + relativedelta(months=+1)).replace(
+                    day=next_day
+                )
```
Idea to avoid the try,

```suggestion
            last_month_day = today + relativedelta(day=31)
            months = 0
            day = 10
            if last_month_day == today:
                months = 1
            else:
                day = {0: 10, 1: 20, 2: 31, 3: 31}.get(today.day // 10)
            next_invoice_date = today + relativedelta(day=day, months=months)
```

lmignon committed:
[FIX] invoicing_ten_days: Ensures one run for same pivot date

Before this change if the cron was launched a nex date at a later time, the invoicing process was run again for the same pivot date.
